xds/src/main/java/io/grpc/xds/XdsServerWrapper.java

```diff
@@ -0,0 +1,439 @@
+      return;
+    }
+    if (isServing) {
+      return;
+    }
+    if (delegate.isShutdown()) {
+      delegate = delegateBuilder.build();
+    }
+    try {
+      delegate.start();
+      listener.onServing();
+      isServing = true;
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(null);
+      }
+    } catch (IOException e) {
+      logger.log(Level.FINE, "Fail to start delegate server: {0}", e);
+      if (!initialStarted) {
+        initialStarted = true;
+        initialStartFuture.set(e);
+      }
+      restartTimer = syncContext.schedule(
+        new RestartTask(), retryDelayNano, TimeUnit.NANOSECONDS, timeService);
+    }
+  }
+
+  private final class RestartTask implements Runnable {
+    @Override
+    public void run() {
+      startDelegateServer();
+    }
+  }
+
+  @VisibleForTesting
+  void setTimeService(@Nullable ScheduledExecutorService scheduledExecutorService) {
+    if (scheduledExecutorService != null) {
+      SharedResourceHolder.release(GrpcUtil.TIMER_SERVICE, timeService);
+    }
+    timeService = scheduledExecutorService;
+  }
+
+  private final class DiscoveryState implements LdsResourceWatcher {
+    private final String resourceName;
+    // Most recently discovered filter chains.
+    private List<FilterChain> filterChains = new ArrayList<>();
+    // Most recently discovered default filter chain.
+    @Nullable
+    private FilterChain defaultFilterChain;
+    private boolean stopped;
+
+    private DiscoveryState(String resourceName) {
+      this.resourceName = checkNotNull(resourceName, "resourceName");
+      xdsClient.watchLdsResource(resourceName, this);
+    }
+
+    @Override
+    public void onChanged(final LdsUpdate update) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          if (update.listener() == null) {
+            // Ignore updates not for servers.
+            return;
+          }
+          filterChains = update.listener().getFilterChains();
+          defaultFilterChain = update.listener().getDefaultFilterChain();
+          updateSelector();
+        }
+      });
+    }
+
+    @Override
+    public void onResourceDoesNotExist(final String resourceName) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          handleConfigNotFound(null);
+          listener.onNotServing(
+              Status.UNAVAILABLE.withDescription(
+                  "Listener " + resourceName + " unavailable").asException());
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Status error) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          if (stopped) {
+            return;
+          }
+          logger.log(Level.FINE, "Transient error from XdsClient: {0}", error);
```

Comment:
why say "transient error" when the very next line has a check for `isPermanentError` ? May be

```
         if (isPermanentError(error)) {
            logger.log(Level.FINE, "Permanent error from XdsClient: {0}", error);
            handleConfigNotFound(null);
          } else {
             logger.log(Level.FINE, "Transient error from XdsClient: {0}", error);
          }
```